lvmd/command/lvm.go

```diff
@@ -504,22 +506,23 @@ func (l *LogicalVolume) Snapshot(name string, cowSize uint64, tags []string, thi
 		// for creating a thin-snapshot, max COW is irrelevant since the snapshot gets resized based on the
 		// thin-pool. Thus one should not define a size argument, otherwise the snapshot is not thin.
 		if !thin {
-			var gbSize uint64
+			var size uint64
 			if cowSize > 0 {
-				gbSize = cowSize >> 30
+				size = cowSize
 			} else {
-				gbSize = (l.size * 2 / 10) >> 30
-				if gbSize > cowMax {
-					gbSize = cowMax
+				// If COW size is not specified, use 20% of the volume size or COW max when it is higher than that.
+				size = l.size * 2 / 10
+				if size > cowMax {
+					size = cowMax
 				}
 			}
-			if gbSize < cowMin {
-				gbSize = cowMin
+			if cowSize < cowMin {
```

Comment:
I doubt that it should be `size` instead of `cowSize`. Could you check it?

Reply:
You are right. However after further review, I am questioning why this code even exists. What is the reason to have a minimum Copy-on-Write Minimum/Maximum? Shouldn't this be irrelevant and we can simply take over the size of the request?

Reply:
The code existed before I joined to maintain it, so I don't know the exact reason. I suspect it is some sort of safeguard to prevent snapshots from becoming unwritable when the CoW space is exhausted.

Reply:
As you know, the code is not called if the volume is thick, we can change the function to only support thin volumes and return an error if the volume is thick.

Reply:
Agreed. Changed the method signature to 

```go
// ThinSnapshot takes a thin snapshot of a volume.
// The volume must be thinly-provisioned.
// snapshots can be created unconditionally.
func (l *LogicalVolume) ThinSnapshot(name string, tags []string) (*LogicalVolume, error)
```

We already error out if the snapshot is not thin at https://github.com/topolvm/topolvm/blob/0e37bda0e8f28c02ac782889047b7c8ede8d47c0/lvmd/lvservice.go#L231-L233 so there is no need for passing thin flag into this method.

Reply:
The naming and the removal of the thin flag are reasonable. However, the LogicalVolume type represents both thin and thick, the ThinSnapshot function can be called even if it is thin. It is better to return an error in this case, IMO.

Reply:
How should we determine if we are a thin snapshot if we do not pass the thin flag? I am not aware how we could deduce that from the given volume. Should we check for attributes first?

Reply:
Removed everything that was executed in this method when `l.pool == nil` because for thin volumes we always have a pool. Also added an error in case we are running without a thin volume

Reply:
LGTM.